contracts/Marketplace.sol

```diff
@@ -1,74 +1,209 @@
+  // ------------------------------------------------------
+
+  function checkServiceOwner(Service memory service) private view {
+    require(service.owner == msg.sender, "Service owner is not the same as the sender");
+  }
+
+  // ------------------------------------------------------
+  // Utils functions
+  // ------------------------------------------------------
 
-    event VersionCreated(bytes32 sid, bytes32 id, bytes32 location, uint256 price);
+  function compare(bytes memory a, bytes memory b) internal pure returns (bool) {
+    return keccak256(a) == keccak256(b);
+  }
 
-    constructor(address creator, bytes32 _sid) Ownable() Pausable() public {
-        sid = _sid;
-        transferOwnership(creator);
-        addPauser(creator);
-        renouncePauser();
+  // ------------------------------------------------------
+  // View functions
+  // ------------------------------------------------------
+
+  // TODO: have to create getter for version and payment because services auto-generated getter doesn't return array of struct
+  function getVersion(uint serviceIndex, uint versionIndex) public view returns (bytes20 hash, bytes memory url) {
+    Version memory version = services[serviceIndex].versions[versionIndex];
+    return (version.hash, version.url);
+  }
+
+  function getPayment(uint serviceIndex, uint paymentIndex) public view returns (address purchaser) {
+    Payment memory payment = services[serviceIndex].payments[paymentIndex];
+    return payment.purchaser;
+  }
+
+  // Index
+
+  function getServiceIndex(bytes memory sid) public view returns (uint serviceIndex) {
+    for (uint i = 0; i < services.length; i++) {
+      if (compare(services[i].sid, sid)) {
+        return i;
+      }
     }
+    require(false, "Service not found");
+  }
 
-    function createVersion(bytes32 versionHash, bytes32 location, uint256 price) public onlyOwner whenNotPaused {
-        require(!versions[versionHash]._exists, "This version already exists");
-        versions[versionHash] = Version({
-            id: versionHash,
-            location: location,
-            price: price,
-            _exists: true
-        });
-        latest = versionHash;
-        emit VersionCreated(sid, versionHash, location, price);
-        return;
+  function getServiceVersionIndex(uint serviceIndex, bytes20 hash) public view returns (uint versionIndex) {
+    Service memory service = services[serviceIndex];
+    for (uint i = 0; i < service.versions.length; i++) {
+      if (service.versions[i].hash == hash) {
+        return i;
+      }
     }
+    require(false, "Version not found");
+  }
 
-    function requestAccess(bytes32 version) public whenNotPaused {
-        require(!_versionAccess[version][msg.sender], "You already have access to this version of the service");
-        _versionAccess[version][msg.sender] = true;
-        return;
+  function getServicePaymentIndex(uint serviceIndex, address purchaser) public view returns (uint paymentIndex) {
```

Comment:
getServicePaymentIndex - why we need this funcrion?

this should be done as external call or removed at all

Reply:
This function is useful because the generated getter of `services` doesn't return payments (and versions).
So it allow the search a payment based on the service index and the address of the purchase instead of having to get every payment from a service to get the right payment.
Maybe this function could return directly the payment instead of its index. It will be even easier.

Reply:
We could extract the payments and versions arrays from Service to put them directly in the state to have the generated getters.

Reply:
And maybe if `Version[] versions` is replace by `uint[] versions` in `Service` struct, the service getter will return it. Need to try.

Reply:
> And maybe if `Version[] versions` is replace by `uint[] versions` in `Service` struct, the service getter will return it. Need to try.

replacing `Version[] versions` by `uint[] versions` doesn't returns versions in getter of `services`.